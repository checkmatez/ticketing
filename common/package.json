{
  "name": "@chm-tickets/common",
  "version": "1.0.20",
  "main": "./build/index.js",
  "types": "./build/index.d.ts",
  "files": [
    "build/**/*"
  ],
  "license": "MIT",
  "scripts": {
    "build": "yarn clean && tsc",
    "clean": "del ./build/*",
    "pub": "git add . && git commit -m \"update\" && yarn publish --new-version patch",
    "prepublishOnly": "yarn build"
  },
  "devDependencies": {
    "del-cli": "^3.0.1",
    "typescript": "^3.9.5"
  },
  "dependencies": {
    "@types/cookie-session": "^2.0.40",
    "@types/express": "^4.17.6",
    "@types/jsonwebtoken": "^8.5.0",
    "cookie-session": "^1.4.0",
    "express": "^4.17.1",
    "express-validator": "^6.5.0",
    "jsonwebtoken": "^8.5.1",
    "node-nats-streaming": "^0.3.2"
  }
}
