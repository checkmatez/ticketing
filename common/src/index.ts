interface Color {
  red: number;
}

export const color: Color = { red: 4 };

console.log(color);
