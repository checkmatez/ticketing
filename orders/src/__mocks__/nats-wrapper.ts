const client = {
  publish: jest.fn().mockImplementation((a, b, c) => c()),
};

export const natsWrapper = {
  getClient: () => client,
};
