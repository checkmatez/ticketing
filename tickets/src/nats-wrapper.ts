import nats, { Stan } from 'node-nats-streaming';

class NatsWrapper {
  private client?: Stan;

  getClient() {
    if (!this.client) {
      throw new Error('NATS client not initialized.');
    }
    return this.client;
  }

  connect(clusterId: string, clientId: string, url: string) {
    this.client = nats.connect(clusterId, clientId, { url });

    return new Promise((resolve, reject) => {
      this.getClient().on('connect', () => {
        console.log('Connected to NATS');
        resolve();
      });
      this.getClient().on('error', (err) => reject(err));
    });
  }
}

export const natsWrapper = new NatsWrapper();
