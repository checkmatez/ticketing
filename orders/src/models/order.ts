import mongoose from 'mongoose';
import { OrderStatus } from '@chm-tickets/common';
import { updateIfCurrentPlugin } from 'mongoose-update-if-current';
import { TicketDoc } from './ticket';

export { OrderStatus };

export interface OrderAttrs {
  userId: string;
  status: OrderStatus;
  expiresAt: Date;
  ticket: TicketDoc;
}

interface OrderModel extends mongoose.Model<OrderDoc> {
  build(attrs: OrderAttrs): OrderDoc;
}

interface OrderDoc extends mongoose.Document {
  userId: string;
  status: OrderStatus;
  expiresAt: Date;
  ticket: TicketDoc;
  version: number;
}

const orderSchema = new mongoose.Schema(
  {
    userId: {
      type: String,
      required: true,
    },
    status: {
      type: String,
      required: true,
      enum: Object.values(OrderStatus),
      default: OrderStatus.Created,
    },
    expiresAt: {
      type: mongoose.Schema.Types.Date,
    },
    ticket: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Ticket',
    },
  },
  {
    toJSON: {
      transform(doc, ret) {
        ret.id = ret._id;
        delete ret._id;
        delete ret.__v;
      },
    },
  },
);

orderSchema.set('versionKey', 'version');

orderSchema.plugin(updateIfCurrentPlugin);

orderSchema.statics.build = (attrs: OrderAttrs) => new Order(attrs);

export const Order = mongoose.model<OrderDoc, OrderModel>('Order', orderSchema);
