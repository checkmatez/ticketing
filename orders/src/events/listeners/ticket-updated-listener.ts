import { Listener, TicketUpdatedEvent, Subject } from '@chm-tickets/common';
import { Message } from 'node-nats-streaming';
import { QUEUE_GROUP_NAME } from './queue-group-name';
import { Ticket } from '../../models/ticket';

export class TicketUpdatedListener extends Listener<TicketUpdatedEvent> {
  readonly subject = Subject.TicketUpdated;

  queueGroupName = QUEUE_GROUP_NAME;

  async onMessage(data: TicketUpdatedEvent['data'], msg: Message) {
    const { id, title, price, version } = data;

    const ticket = await Ticket.findByEvent(data);
    if (!ticket) {
      throw new Error('Ticket not found');
    }

    ticket.set({ title, price });
    await ticket.save();

    msg.ack();
  }
}
