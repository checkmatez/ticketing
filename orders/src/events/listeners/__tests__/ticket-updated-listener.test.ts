import { TicketUpdatedEvent } from '@chm-tickets/common'
import { Ticket } from '../../../models/ticket'
import { natsWrapper } from '../../../nats-wrapper'
import { TicketUpdatedListener } from '../ticket-updated-listener'

const fakeId = '5efb8a95b3b9995358d25e96'

const setup = async () => {
  const listener = new TicketUpdatedListener(natsWrapper.getClient())

  const ticket = Ticket.build({
    id: fakeId,
    price: 20,
    title: 'concert',
  })
  await ticket.save()

  const data: TicketUpdatedEvent['data'] = {
    id: ticket.id,
    version: ticket.version + 1,
    title: 'new concert',
    price: 30,
    userId: 'asd1',
  }

  const msg = {
    ack: jest.fn(),
  }

  return { listener, data, msg, ticket }
}

it('finds, updates and saves a ticket', async () => {
  const { listener, data, msg, ticket } = await setup()

  await listener.onMessage(data, msg as any)

  await expect(Ticket.findById(data.id)).resolves.toMatchSnapshot()
})

it('acks the message', async () => {
  const { listener, data, msg } = await setup()
  await listener.onMessage(data, msg as any)

  expect(msg.ack).toBeCalled()
})

it('does not call ack if the event skipped version number', async () => {
  const { listener, data, msg } = await setup()
  data.version = 10

  await expect(listener.onMessage(data, msg as any)).rejects.toThrow()

  expect(msg.ack).not.toBeCalled()
})
