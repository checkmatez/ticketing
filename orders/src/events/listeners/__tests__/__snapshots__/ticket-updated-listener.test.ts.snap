// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finds, updates and saves a ticket 1`] = `
Object {
  "id": "5efb8a95b3b9995358d25e96",
  "price": 30,
  "title": "new concert",
  "version": 1,
}
`;
